src/traits.jl

```diff
@@ -514,14 +514,31 @@ isapprox(x::AbstractGray, y::Number; kwargs...) =
 
 
 zero(::Type{C}) where {C<:Gray} = C(0)
+zero(::C) where {C<:Gray} = C(0)
 oneunit(::Type{C}) where {C<:Gray} = C(1)
+oneunit(::C) where {C<:Gray} = C(1)
 
 function Base.one(::Type{C}) where {C<:Gray}
     Base.depwarn("one($C) will soon switch to returning 1; you might need to switch to `oneunit`", :one)
     C(1)
 end
+Base.one(::C) where {C<:Gray} = one(C)
 
 Base.broadcastable(x::Colorant) = Ref(x)
 
+# specialize ones and zeros for Gray type:
+#   * Base implementation uses `one` instead of `oneunit`; however, `one(Gray)` will return `1`
+#   * Base implementation use `Array{T}` to initialize the array, however, if we pass
+#     `Gray`, it gives an abstract eltype `Array{Gray}`, which is inconsistent to `zero(Gray)`
```

Comment:
To be honest, I don't understand what this comment says. I feel the reason is the opposite.
Even though `one(Gray)` should return `1`, not so. Thus, `one` shows the warning. Isn't this PR a workaround to avoid that warning? If `one(Gray)` returns `1`, ones(Gray) returns an array of Gray(1).

Regarding the latter:
```julia
julia> zero(AbstractFloat)
0.0

julia> zeros(AbstractFloat)
0-dimensional Array{AbstractFloat,0}:
0.0
```
`Gray{N0f8}` is just a matter of practicality, i.e. speed or memory efficiency.

Reply:
I meant to make `ones(Gray)` still return an array of `Gray(1)` regardless of the fact that `one(Gray) === 1` in the future. Did I misunderstand the original reason for the shift?

I think I need to raise `zeros(AbstractFloat)` issue at julialang/julia.. My idea is to make the array eltype equals to its initializer, which makes more sense to me.

Reply:
> I think I need to raise zeros(AbstractFloat) issue at julialang/julia..

Umm... What is the problem?

Reply:
That `eltype(ones(AbstractFloat)) === typeof(one(AbstractFloat))` (and for `zeros`)

Reply:
Yes. I understand the claim, but I just want to know the reason.

Reply:
Just opened a PR for this https://github.com/JuliaLang/julia/pull/34948

wrt whether `ones(Gray{N0f8}, 4, 4)` should return array of `N0f8` or `Gray{N0f8}`, I'd like to get @timholy involved.

Reply:
Definitely, `ones(Gray{N0f8}, 4, 4)` should not return an array of `N0f8`. Where did the choice come from?

```julia
ones(Gray{N0f8}, 2, 2)
--> Gray{N0f8}[? ?
               ? ?]
--> Gray{N0f8}[one(Gray{N0f8}) one(Gray{N0f8})
               one(Gray{N0f8}) one(Gray{N0f8})]
--> Gray{N0f8}[1 1
               1 1]
--> Gray{N0f8}[Gray{N0f8}(1) Gray{N0f8}(1) 
               Gray{N0f8}(1) Gray{N0f8}(1)]
```